src/sorting/quicksort.jl

```diff
@@ -0,0 +1,416 @@
+        end
+    end
+    CUDA.device_synchronize()
+    sync_threads()
+    #= step 3: consolidate the partitioned batches so that the sublist from
+    [lo, hi) is partitioned, and the partition is stored in `partition`=#
+
+    partition = -1
+    if L <= N_t ^ 2
+        #= if small enough, perform binary search to find the partitions in each
+        batch, store in shmem, and use that=#
+        init_bsums(vals, pivot, lo, L, N_t, N_b, i, b_sums, P)
+        partition = consolidate_batch_partition(vals, pivot, lo, L, N_t, N_b, i, b_sums)
+    else
+        #= if more batches than threads per SM, use the global array written to
+        by `partition_batches_kernel`=#
+        partition = consolidate_batch_partition(vals, pivot, lo, L, N_t, N_b, i, global_b_sums)
+    end
+
+    #= step 4: recursion. If the first subproblem doesn't require global mem,
+    run it on a temporary stream. Otherwise run in default stream, so that
+    the two subproblems run serially, and don't conflict over the global array=#
+    if i == 1 && partition > lo
+        if partition - lo <= N_t ^ 2
+            s = CuDeviceStream()
+            @cuda threads=N_t dynamic=true stream=s qsort_kernel(vals, lo, Int32(partition), global_b_sums, block_dim, Val(!P), depth + Int8(1))
+            CUDA.unsafe_destroy!(s)
+        else
+            @cuda threads=N_t dynamic=true qsort_kernel(vals, lo, Int32(partition), global_b_sums, block_dim, Val(!P), depth + Int8(1))
+        end
+    end
+
+    if i == 1 && partition < hi
+        if hi - partition <= N_t ^ 2
+            s = CuDeviceStream()
+            @cuda threads=N_t dynamic=true stream=s qsort_kernel(vals, Int32(partition), hi, global_b_sums, block_dim, Val(!P), depth + Int8(1))
+            CUDA.unsafe_destroy!(s)
+        else
+            @cuda threads=N_t dynamic=true qsort_kernel(vals, Int32(partition), hi, global_b_sums, block_dim, Val(!P), depth + Int8(1))
+        end
+    end
+    return nothing
+end
+
+function quicksort!(c :: CuArray{T}, block_size :: Integer) where T
+    @assert isinteger(Base.log2(block_size)) "`block_size` must be a power of two"
+    CUDA.limit!(CUDA.LIMIT_DEV_RUNTIME_SYNC_DEPTH, 24)
+    N = length(c)
+    b_sums = CuArray{Int32}(undef, ceil_div(N, block_size))
+    CUDA.@sync(@cuda threads=block_size qsort_kernel(c, Int32(0), N, b_sums, Val(Int32(block_size)), Val(true), Int8(1)))
+end
+
+"""
+Attempts to determine the maximum blocksize for running quicksort on the
+available hardware. Picks the greatest power of two which doesn't violate:
+    - Inherent maximum block size
+    - Register usage
+    - Shared memory
+This assumes that register usage doubles with elevated debug level, but usage of
+other resources is unaffected.
+"""
+function quicksort_block_size(T)
+    # requirements
+    # this is a guess for older architectures. 64 is fine for RTX 2070
+    registers_per_thread = 128
+    if Base.JLOptions().debug_level >= 2
+        registers_per_thread *= 2
```

Comment:
That's pretty inaccurate. You can always inspect the register usage of a compiled kernel. But even better, can't you use the occupancy API and skip inspecting device attributes below, too?

Reply:
I didn't know that functionality existed - that's awesome! I had just measured register usage with nvvp locally. I see how I could set a `config` option when I call `@cuda`, and that could set threads based on register usage (or via the functions in `occupancy.jl`). There is a slight problem with that. `qsort_kernel` sets static shmem using a Val type, and that Val matches block size. I don't see an easy way to use the `config` option unless I switch to dynamic shmem, which seems unnecessary.

One way I could see getting around this is creating a macro which finds the largest block size possible (out of [256, 512, 1024, 2048]). This could be done by calling creating a `HostKernel` for each block size and using the occupancy API to pick the largest viable option. Each `HostKernel` would have the corresponding Val argument. Does that seem like a reasonable approach or is there something easier?

Reply:
> `qsort_kernel` sets static shmem using a Val type, and that Val matches block size. I don't see an easy way to use the `config` option unless I switch to dynamic shmem, which seems unnecessary.

Why don't you want to use dynamic shared memory? It's exactly meant for this use case. Otherwise you'd compile multiple kernels, and the occupancy API supports using a closure that determines the amount of dynamic shared memory required for this configuration.

Reply:
The original reason was performance. There was something like a factor of two difference for `partition_batches_kernel` (as profiled by nvvp).  I never committed that code unfortunately. Assuming performance is significantly better, having a few extra methods sounds like a worthwhile cost, no?

The real, current reason is if I switch to using dynamic memory the code breaks. `partition_batches_kernel` seems to get confused because it uses two arrays in shared memory. When using static shmem, it works fine.

Either this is a bug, or I misunderstood how to allocate dynamic shared memory. I've taken the risk of filing an issue :)
https://github.com/JuliaGPU/CUDA.jl/issues/555

Reply:
> The original reason was performance.

I don't understand this; there shouldn't be a difference, unless you use different amounts of memory. Are you sure your `shmem` calculation was correct? It should be in bytes, not in number of elements.

Reply:
Ok - thanks for pointing out the offset logic. I can't seem to reproduce the performance gap which is good. The overall sort runs in about the same amount of time. Interestingly there is a big jump in register usage: 48 -> 72.